web/services/premium_detection.go

```diff
@@ -0,0 +1,72 @@
+package services
+
+import (
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	Premium   = "Premium"
+	Community = "Community"
+)
+
+//go:generate mockery --name=PremiumDetection --inpackage --filename=premium_detection_mock.go
+
+type PremiumDetectionService interface {
+	RequiresEulaAcceptance() (bool, error)
+	CanPublishTelemetry() (bool, error)
+	IsPremiumActive() (bool, error)
+}
+
+type premiumDetection struct {
+	flavor        string
+	subscriptions SubscriptionsService
+	settings      SettingsService
+}
+
+func NewPremiumDetection(flavor string, subscriptions SubscriptionsService, settings SettingsService) *premiumDetection {
+	return &premiumDetection{
+		flavor,
+		subscriptions,
+		settings,
+	}
+}
+
+func (premiumDetection *premiumDetection) RequiresEulaAcceptance() (bool, error) {
+	if !premiumDetection.isPremiumFlavor() {
+		return false, nil
+	}
+	isEulaAccepted, err := premiumDetection.settings.IsEulaAccepted()
+	if err != nil {
+		log.Errorf("Unable to determine whether the EULA has been accepted. Error: %s", err)
+		return false, err
+	}
+	return !isEulaAccepted, err
+}
+
+func (premiumDetection *premiumDetection) CanPublishTelemetry() (bool, error) {
+	if !premiumDetection.isPremiumFlavor() {
+		return false, nil
+	}
+	isEulaAccepted, err := premiumDetection.settings.IsEulaAccepted()
+	if err != nil {
+		log.Errorf("Unable to determine whether telemetry can be published. Error: %s", err)
+		return false, err
+	}
+	return isEulaAccepted, nil
+}
```

Comment:
Can't help but noticing that these two methods are basically identical except a `not` operator, but hey.

Reply:
They are expressing two different domain invariants, and a little bit of duplication in favour of explicitness I find it acceptable.

Reply:
fair enough 😉

Reply:
I'd keep the whole function in `RequiresEulaAcceptance` and do `return !premiumDetection.RequiresEulaAcceptance` in `CanPublishTelemetry`

Reply:
nah, Nelson is right, these are two different domain concepts, if we couple them together it's gonna bite us back eventually, I was hinting at maybe find a common domain invariant, but it's good enough as it is.